Fix OSV to handle affected_packages correctly & add support to collect commits

ziadhany opened the pull request:
Issue:
 - #2066
- #2091

ziadhany commented:
OSV Logs , importers:
- pypa_importer_v2
- pysec_importer_v2
- oss_fuzz_importer_v2
- github_osv_importer_v2


[osv_logs.zip](https://github.com/user-attachments/files/24328393/osv_logs.zip)


Also , There is still a small bug regarding the invalid constraints sequence, I’m trying to find a solution for it.
ex:
```
univers.version_constraint.InvalidConstraintsError: Invalid constraints sequence: ('>= 2.6','>= 2.7') in [VersionConstraint(comparator='<', version=PypiVersion(string='2.5.11')), VersionConstraint(comparator='>=', version=PypiVersion(string='2.6')), VersionConstraint(comparator='>=', version=PypiVersion(string='2.7'))]
```

ziadhany commented:
For logs, we need to make a clear decision about which logs and data we ignore, and which ones we need to fix/support.

pysec_importer_v2:

```
ERROR 2025-12-31 05:03:11.688415 UTC Unsupported PyPI advisory data file: GHSA-227r-w5j2-6243.json
```
```
ERROR 2025-12-31 05:03:12.411085 UTC Unsupported PyPI advisory data file: MAL-2025-917.json
```

oss_fuzz_importer_v2

```
Unsupported package type: {'package': {'name': 'libbpf', 'ecosystem': 'OSS-Fuzz'}, 'ranges': [{'type': 'GIT', 'repo': 'https://github.com/libbpf/libbpf', 'events': [{'introduced': '421213a052aebb0c357b6d0872d6c57f2113800d'}, {'fixed': '741277511035893c72a34df05da3b943afa747a4'}]}], 'versions': ['v0.6.0', 'v0.6.1', 'v0.7.0', 'v0.8.0', 'v0.8.1', 'v1.0.0', 'v1.0.1'], 'ecosystem_specific': {'severity': 'HIGH'}} in OSV: 'OSV-2021-1576'
```


github_osv_importer_v2

```
Unsupported package type: {'package': {'ecosystem': 'Go', 'name': 'github.com/cortexproject/cortex'}, 'ranges': [{'type': 'ECOSYSTEM', 'events': [{'introduced': '0'}, {'last_affected': '1.9.0'}]}]} in OSV: 'GHSA-jphm-g89m-v42p'
```


```
Unsupported package type: {'package': {'ecosystem': 'crates.io', 'name': 'deno'}, 'ranges': [{'type': 'ECOSYSTEM', 'events': [{'introduced': '1.5.0'}, {'fixed': '1.10.2'}]}], 'database_specific': {'last_known_affected_version_range': '<= 1.10.1'}} in OSV: 'GHSA-xpwj-7v8q-mcgj'
```


```
Unsupported package type: {'package': {'ecosystem': 'Pub', 'name': 'archive'}, 'ranges': [{'type': 'ECOSYSTEM', 'events': [{'introduced': '0'}, {'fixed': '3.3.8'}]}], 'database_specific': {'last_known_affected_version_range': '<= 3.3.7'}} in OSV: 'GHSA-r285-q736-9v95'
```

```
Unsupported package type: {'package': {'ecosystem': 'SwiftURL', 'name': 'github.com/marmelroy/Zip'}, 'ranges': [{'type': 'ECOSYSTEM', 'events': [{'introduced': '0'}, {'last_affected': '2.1.2'}]}]} in OSV: 'GHSA-g454-wj9r-jpg4'
```

keshav-space commented on vulnerabilities/tests/test_data/osv_test/github/github-expected-3.json:
Here also we should get only one affected pacakge for a single range.

keshav-space reviewed (changes requested):
Thanks @ziadhany, see some comments below.

Regarding `pysec_importer_v2` log it is very vague. It gives no information about what the problem with the advisory. Is it that we can’t parse the version, package is missing, or something else?  Log should be actionable.

ziadhany commented:
@keshav-space The problem is that in OSV we have **three types of version ranges**:

- Explicit version list

```
"versions": [
  "2.8.0", "2.8.0.post1", "2.8.0.post2",
  "2.9.0", "2.9.1", "2.9.2"
]
```

- Range-based version

```
{
  "type": "ECOSYSTEM",
  "events": [
    { "introduced": "2.8.0" },
    { "fixed": "2.10.0" }
  ]
}
```

- `last_known_affected_version_range`

```
"database_specific": {
  "last_known_affected_version_range": "<= 0.21.7"
}
```

We **cannot merge these**, because doing so would produce an invalid version constraint.

For example:
[https://github.com/github/advisory-database/blob/8a993f4259457ec9f447777c053316da28eb5908/advisories/github-reviewed/2020/08/GHSA-wwrj-35w6-77ff/GHSA-wwrj-35w6-77ff.json#L9](https://github.com/github/advisory-database/blob/8a993f4259457ec9f447777c053316da28eb5908/advisories/github-reviewed/2020/08/GHSA-wwrj-35w6-77ff/GHSA-wwrj-35w6-77ff.json#L9)

This results in the following error:

```
univers.version_constraint.InvalidConstraintsError:
Invalid constraints sequence: ('< 1.0.5', '< 1.0.7')
in [
  VersionConstraint(comparator='<', version=SemverVersion(string='1.0.5')),
  VersionConstraint(comparator='<', version=SemverVersion(string='1.0.7'))
]
```

ziadhany commented:
@keshav-space I updated OSV to prioritize the affected explicit version list first, then last_known_affected_version_range, and finally the range-based versions, removing duplicates. I also updated the OSV logs.

For PySec logs:
Some files are not supported right now, for example:

- `MAL-2023-1783.json`  - #2019

Some files are intentionally ignored. When we wrote this importer, we decided to skip such files and focus only on the PYSEC- files, for example:
- `GHSA-227r-w5j2-6243.json`
